feat: Checkbox Component 완료

mkachi opened the pull request:
feat: Checkbox Component 작업 완료

mkachi committed:
feat: Checkbox Component 완료

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
````diff
@@ -0,0 +1,194 @@
+ * @param {CheckboxProps} {@link CheckboxProps} 기본적인 Checkbox 요소
+ *
+ * @example ```tsx
+ * <Checkbox checked={checked} onChange={onChange}>Checkbox</Checkbox>
````
```tsx
const [checked, setChecked] = useState(false)
<Checkbox checked={checked} onChange={() => setChecked((state) => !state)}>CheckBox</Checkbox>
```

mkachi replied:
사용 예시를 다시 작성하도록 하겠습니다

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
````diff
@@ -0,0 +1,194 @@
+ * ```
+ */
+export const Checkbox: FC<Partial<CheckboxProps>> = (props) => {
+  const elementProps = useCheckbox(props)
````
```tsx
const {
  ...
} = useCheckbox(props)
```

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
````diff
@@ -0,0 +1,194 @@
+ * <Checkbox checked={checked} onChange={onChange}>Checkbox</Checkbox>
+ * ```
+ */
+export const Checkbox: FC<Partial<CheckboxProps>> = (props) => {
````
https://ko.reactjs.org/docs/forwarding-refs.html

Vallista replied:
```tsx
const text = React.forwardRef<HTMLInputElement, { a: string }>((props, ref) => {
  return (
    <input ref={ref}>
      123
    </div>
  )
})
```

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+  } = elementProps
+
+  return (
+    <Label
```
```tsx
const { onCheckboxClick, ...otherProps } = useCheckbox(props)

<Label {...otherProps} onClick={() => onCheckboxClick(ref)}>
</Label>
```

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+      onMouseEnter={onMouseEnter}
+      onMouseLeave={onMouseLeave}
+      style={style}
+      onClick={() => onCheckboxClick(ref)}
```
```tsx
(e) => onCheckboxClick(e)
```

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+      {label && <LabelText>{label}</LabelText>}
+      <Input
+        ref={ref}
+        type={'checkbox'}
```
`type='checkbox'` = O
`type={'checkbox'}` = X

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+    css`
+      align-items: flex-start;
+      flex-direction: column;
+      display: inline-flex;
```
순서

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+}
+
+const Label = styled.label<Pick<ReturningUseCheckbox, 'fullWidth' | 'disabled' | 'label'>>`
+  font-size: 0.875rem;
```
em, rem 공부
root html default font-size = ?

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+  `}
+`
+
+const LabelText = styled.div`
```
다시 재고

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+  margin-bottom: 8px;
+`
+
+const Input = styled.input`
```
컬링 안됨

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+`
+
+const Input = styled.input`
+  position: absolute;
```
top, left 0

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+  ${({ isFocus, theme }) => css`
+    ${isFocus &&
+    css`
+      box-shadow: 0 0 0 2px ${theme.colors.PRIMARY.BACKGROUND}, 0 0 0 4px ${theme.colors.PRIMARY.ACCENT_3};
```
theme.shadows

Vallista commented on packages/core/src/components/Checkbox/type.ts:
```diff
@@ -0,0 +1,25 @@
+  label: string
+  children: ReactNode
+  style: CSSProperties
+  onChange(event: React.ChangeEvent<HTMLInputElement>): void
```
- `onChange(event: ChangeEvent<HTMLInputElement>): void`
- `onChange: (e: ChangeEvent<HTMlInputElement>) => void`

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,194 @@
+      box-shadow: 0 0 0 2px ${theme.colors.PRIMARY.BACKGROUND}, 0 0 0 4px ${theme.colors.PRIMARY.ACCENT_3};
+    `}
+  `}
+
```
```css
input:hover + div {
  ...
}
```

Vallista commented on packages/core/src/components/Checkbox/useCheckbox.ts:
```diff
@@ -0,0 +1,69 @@
+
+export const useCheckbox = <T extends Partial<CheckboxProps>>(props: T): ReturningUseCheckbox => {
+  const [focus, setFocus] = useState(false)
+  const [hover, setHover] = useState(false)
```
필요없음

Vallista commented on packages/core/src/components/Checkbox/useCheckbox.ts:
```diff
@@ -0,0 +1,69 @@
+  return {
+    ...initProps,
+    ...props,
+    marker,
```
marker: props.indeterminate ? 'indeterminate' : props.checked ? 'checked' : 'none'

Vallista reviewed (commented):
ㅁㄴㅇ

mkachi committed:
fix: checkbox @example 주석 수정

mkachi committed:
fix: 1차 리뷰 반영

fix: focus, hover 상태 구분을 CSS 선택자로 구분하도록 수정
fix: 필요 없는 ref 제거
fix: 각 Element들의 property를 spread 연산자를 통해 전달하도록 수정
fix: onChange의 타입 선언을 method에서 property로 변경

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,165 @@
+  position: absolute;
+  top: 0;
+  left: 0;
+  opacity: 0;
```
얜 안보여야하니까 left, top 에 -9999 등을 넣어줘야할 듯, z-index도 9999
theme layer 참고

Vallista commented on packages/core/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,165 @@
+`
+
+const Box = styled.div<Pick<ReturningUseCheckbox, 'checked' | 'disabled' | 'indeterminate'>>`
+  width: 1em;
```
왜 1em이죠

mkachi committed:
fix: 화면에서 보이지 않도록 top, left, z-index 수정

mkachi committed:
fix: Checkbox의 크기가 항상 일정하도록 width, height를 고정 pixel 값으로 수정

mkachi committed:
feat: CheckboxPlayground 추가

Vallista commented on packages/core/src/components/Checkbox/useCheckbox.ts:
```diff
@@ -0,0 +1,22 @@
+  indeterminate: false,
+  disabled: false,
+  fullWidth: false,
+  onChange: (): void => {
```
onChange: () => void 0

mkachi committed:
fix: 3차 리뷰 반영

fix: checkbox 의 onChange 기본값 수정

mkachi committed:
feat: checkbox playground 추가
